### PR TITLE
Revert "librbd/cache/pwl: fix bug of flush request blocked by deferd IO"

### PR DESCRIPTION
**Pull request by CongMinYin**
This reverts commit 6583182.
It cause that flush requests are dispatched in advance, which
may lead to semantic changes of user requests and data loss.

Fixes: https://tracker.ceph.com/issues/52599

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by CongMinYin** [comment]
I am very sorry for the fatal problem caused by the PR-40208. I should bear the responsibility. At that time, I had just joined the project. When I was not familiar with the program, This PR-40208 caused many big problem in order to solve a small problem. Since the former authors have left and maintainer is about to leave, the review merging process is not rigorous. I found this problem when I reviewed the historical commit again to solve the problem of too long write back time.

**Comment by CongMinYin** [comment]
This may also be the root cause of the SSD sync point problem(long stack), which is being verified.

**Comment by CongMinYin** [comment]
@idryomov Please process this PR with higher priority.  And CC @majianpeng  @hualongfeng

**Comment by MahatiC** [comment]
> I am very sorry for the fatal problem caused by the PR-40208. I should bear the responsibility. At that time, I had just joined the project. When I was not familiar with the RWL, This PR-40208 caused many big problem in order to solve a small problem. Since the former authors have left and maintainer is about to leave, the review merging process is not rigorous. I found this problem when I reviewed the historical commit again to solve the problem of too long write back time.

Hi @CongMinYin please don't worry about it. As you rightly pointed, things have been a bit out of sync on our end for various reasons.

Could you point to the crash (a tracker ticket if there is one) that was happening due to this PR? This is to verify that this PR is indeed the cause. Also, if there is no tracker for this - could you create one?

**Comment by CongMinYin** [comment]
> Could you point to the crash (a tracker ticket if there is one) that was happening due to this PR? This is to verify that this PR is indeed the cause. Also, if there is no tracker for this - could you create one?

The most obvious problem is that after the wrong PR is applied, the shutdown phase will last a very long write back time, because the sync point is not inserted in the correct position and no flush request to write data back to osd. I can create a tracker for this. For example, F mean flush req, W mean Write req.
The right sequence is:
F->W->W->W>...->W(256times)->F->W->W->W>...->W(256times)->F->W->W->W>...->W(256times)->F
Now it becomes like:
F->W->F->W->W->F->W>F->...->W(many data entries occupies the whole cache space and is not written back in time)

In addition, executing the flush request in advance will cause sync point problems, and there is already a tracker(https://tracker.ceph.com/issues/51418). Some validation has been carried out this afternoon and further validation is needed.

**Comment by CongMinYin** [comment]
create tracker: https://tracker.ceph.com/issues/52258 and update commit message

**Comment by CongMinYin** [comment]
https://github.com/ceph/ceph/pull/40208 uses the wrong method to solve the problem. It is going to be reverted, so I create a tracker for the original problem, giving the reproduction method and the expected solution. https://tracker.ceph.com/issues/52277

From the test results, this PR https://github.com/ceph/ceph/pull/40208 will cause the write back phase to last for a long time, and also cause the problem of sync point issue(https://tracker.ceph.com/issues/51418), But I still don't know how the code logic causes these unexpected errors. This can be in the follow-up study. To avoid these unknown errors, revert is the best way now.

**Comment by ideepika** [comment]
https://pulpito.ceph.com/ideepika-2021-08-15_00:22:12-rbd-wip-deepika-testing-2021-08-13-2138-distro-basic-smithi
rerun: https://pulpito.ceph.com/ideepika-2021-08-16_08:32:04-rbd-wip-deepika-testing-2021-08-13-2138-distro-basic-smithi

File: src/librbd/cache/pwl/AbstractWriteLog.cc
Comment (idryomov) [inline src/librbd/cache/pwl/AbstractWriteLog.cc:834]: Is this comment really outdated?  IIUC you are still planning to switch from libpmemobj to raw libpmem and therefore manual allocation management?

Comment (CongMinYin) [reply src/librbd/cache/pwl/AbstractWriteLog.cc:834]: > Is this comment really outdated? IIUC you are still planning to switch from libpmemobj to raw libpmem and therefore manual allocation management (#41432)?

I think it is outdated. I have the plan to manage pmem space by libpmem. But PMDK libpmemobj don't have fragment problems now. It fixed in https://github.com/ceph/ceph/pull/40493 by new method. Therefore, in fact, the split operation is not necessary, but in order to make it easier to allocate space by libpmemobj and reduce the defragmentation operation, this step is still beneficial. If you think it necessary to explain, the comment should be changed to "split blocks to make it easier to allocate space, reduce pmem's defragmentation operations, and plan to use libpmem to manage space by itself in the future"

Comment (idryomov) [reply src/librbd/cache/pwl/AbstractWriteLog.cc:834]: How about
```
  /*
   * Split image extents larger than 1M.  This isn't strictly
   * necessary but makes libpmemobj allocator's job easier and
   * reduces pmemobj_defrag() cost.  We plan to manage pmem space
   * and allocation by ourselves in the future.
   */
```

Comment (CongMinYin) [reply src/librbd/cache/pwl/AbstractWriteLog.cc:834]: done

**Comment by idryomov** [comment]
Could you provide an example of a workload where the shutdown phase timing changed significantly?  Teuthology tests have passed (no new failures, that is) but in my local tests I don't see much of a difference.  I tried only the ssd mode so far though...

**Comment by CongMinYin** [comment]
> Could you provide an example of a workload where the shutdown phase timing changed significantly? Teuthology tests have passed (no new failures, that is) but in my local tests I don't see much of a difference. I tried only the ssd mode so far though...

I have just do many tests in RWL mode after you asked the workload, and the results are different from those I have obtained before. I may be cheated by some fluctuating test results. With or without this PR, the exit time of RBD bench is quite long in many cases. In my environment, 1GB image takes about 20s, and it takes nearly 10s to exit. The time of FIO test from the appearance of test results to exit seems to fluctuate greatly. Sometimes it exits quickly, and sometimes it takes 20 ~ 30s. I apologize for my incomplete test before. On the issue of sync piont, I asked Hualong to help me do the test. He tested three times in the old version and bug did not appear. Because it is not inevitable, it may also be a coincidence.

Tomorrow I will do more test and print the amount of dirty entries directly during shutdown to verify. If this PR wastes your time, please forgive me.

**Comment by idryomov** [comment]
> If this PR wastes your time, please forgive me.

Not at all!  Just want to get to the bottom of this because the shutdown taking too long seems to affect both rwl and ssd modes and shows up in benchmarks.

**Comment by CongMinYin** [comment]
If the bandwidth of write to cache is larger than that of write back to cluster(general case):

cache not full: fill_in_cache_time = cache_cap / (bw_write_to_cache - bw_write_back_to_cluster)

cache full, Stable_stage: bw_write_to_cache speed will decrease to bw_write_back_to_cluster

write end: write back time = cache_cap / bw_write_back_to_cluster

The time of writing back to the cluster is the time saved by high-speed writing when the cache has free space. For the time being, this is inevitable and logical. Writing back to the cluster is faster than writing directly to the cluster when the thread or depth is not high.

**Comment by CongMinYin** [comment]
@idryomov I still want to reverse this pr. It's not because of the performance problem,   the user semantics has changed. If the user sends **write1 write2 write3 write4 flush write5**, and the depth is 2, it may become **write1 write2 write3 flush write4 write5**. Although this will not cause errors, the semantics has changed from 1234 | 5 to 123 | 45. This may cause unpredictable errors when there is overlap, but QA test can't find out. Should I create a new PR or reopen this PR？

**Comment by idryomov** [comment]
If "librbd/cache/pwl: fix bug of flush request blocked by deferd IO" commit can indeed cause "1234 | 5" to become "123 | 45" it definitely needs to be reverted.  Moving the flush to the right (e.g. "12 | 345" becoming "1234 | 5") is fine, moving the flush to the left (your example) is a potential data loss bug.

Please reopen this PR.

**Comment by idryomov** [comment]
I think we also need a new tracker ticket since the revert won't fix https://tracker.ceph.com/issues/52258.

**Comment by CongMinYin** [comment]
> I think we also need a new tracker ticket since the revert won't address https://tracker.ceph.com/issues/52258 which is currently linked in the description

Please check. Do I need to give examples in the description, or is such a simple explanation enough?

**Comment by idryomov** [comment]
Looks good!

**Review by idryomov — DISMISSED** [review]

**Comment by idryomov** [comment]
This seems to result in a hang while running xfstests, both for rwl and ssd.  I'm not sure if it stalls completely or just becomes very slow but test jobs time out after 12 hours.  Yet to confirm by rerunning without this PR.

**Comment by CongMinYin** [comment]
> This seems to result in a hang while running xfstests, both for rwl and ssd. I'm not sure if it stalls completely or just becomes very slow but test jobs time out after 12 hours. Yet to confirm by rerunning without this PR.

It has been tested before and there is no problem, but I rebased this branch several days ago. Could you please share the log link. 
I will also run qa tests on my teuthology platform, try to reproduce.

**Comment by idryomov** [comment]
master at f110ec33e5c5eef3fb6101780f84c046ccb6fc99:

https://pulpito.ceph.com/dis-2021-09-16_21:59:41-rbd-wip-dis-testing-1-distro-basic-smithi/

master at f110ec33e5c5eef3fb6101780f84c046ccb6fc99 + this PR:

https://pulpito.ceph.com/dis-2021-09-16_09:21:06-rbd-wip-dis-testing-distro-basic-smithi/

So definitely down to this PR.  It hangs while executing test 074.  `tp_pwl` threads are doing work but no progress is being made:

```
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  16427 ubuntu    20   0    9.8g   6.4g  42536 S  46.3  20.4 132:36.54 tp_pwl
  16428 ubuntu    20   0    9.8g   6.4g  42536 R  45.7  20.4 125:53.02 tp_pwl
  16429 ubuntu    20   0    9.8g   6.4g  42536 R  44.3  20.4 125:27.71 tp_pwl
  16430 ubuntu    20   0    9.8g   6.4g  42536 S  41.0  20.4 125:51.25 tp_pwl
```

**Comment by CongMinYin** [comment]
@idryomov the new commit can fix 074 issue. This is the bug solved by revert's PR. now I solved it again by using another method. But after my test, sometimes there is a output mismatch issue between 086 or 259, like - `output mismatch (see 086.out.bad)` . This is not caused by this pr, it often occurs in qa of ssd mode(master), and rwl also occasionally occurs after the PR is applied. This should be an error at some time due to different performance. I suggest holding this PR until `output mismatch` fixed. What do you think?

**Comment by majianpeng** [comment]
@idryomov . in our xfstest,  We often encounter 259 failed.  I compiled kernel to find the root cause. I found this issue is kernel loop driver bug. the link is http://lkml.iu.edu/hypermail/linux/kernel/1209.3/02761.html.  I think librbd cache increase the probability of this problem．
BTW: why we use so old kernel(ubuntu 12.02, kernel 3.2.0-32)?

**Comment by majianpeng** [comment]
@CongMinYin  I reviewed code again and thought we don't need revert the commit: https://github.com/ceph/ceph/commit/6583182f624386bfbb7375fb227dfb77b0a3209d. 
commit https://github.com/ceph/ceph/commit/6583182f624386bfbb7375fb227dfb77b0a3209d make flush_req dispatch in advace in func alloc_and_dispatch_io_req.
But flush_req is a special req. When dose the flush_req call alloc_and_disptach_io_req?
flush_rq is belong to SyncPoint. Only all ops of syncpoint already appened to m_ops_to_append queue, it can call alloc_and_dispatch_io_req. So currently, m_deferred_ios isn't empty, the deferred ops belonged  to new_sync_point. 
Although flush_req dispatch in advance but it don't change the semantics  which flush previous ops rather than later ops.
So i think we shouldn't revert commit  https://github.com/ceph/ceph/commit/6583182f624386bfbb7375fb227dfb77b0a3209d.

**Comment by CongMinYin** [comment]
> @CongMinYin I reviewed code again and thought we don't need revert the commit: [6583182](https://github.com/ceph/ceph/commit/6583182f624386bfbb7375fb227dfb77b0a3209d). commit [6583182](https://github.com/ceph/ceph/commit/6583182f624386bfbb7375fb227dfb77b0a3209d) make flush_req dispatch in advace in func alloc_and_dispatch_io_req. But flush_req is a special req. When dose the flush_req call alloc_and_disptach_io_req? flush_rq is belong to SyncPoint. Only all ops of syncpoint already appened to m_ops_to_append queue, it can call alloc_and_dispatch_io_req. So currently, m_deferred_ios isn't empty, the deferred ops belonged to new_sync_point. Although flush_req dispatch in advance but it don't change the semantics which flush previous ops rather than later ops. So i think we shouldn't revert commit [6583182](https://github.com/ceph/ceph/commit/6583182f624386bfbb7375fb227dfb77b0a3209d).

For external flush request, it new sync point directly before dispatch.  However, the last write request may still be in the deffered queue, It don't dispathc and  not associated with syncpoint. So external flush is different from internal flush. It should strictly follow the order of dispatch.

**Comment by majianpeng** [comment]
@CongMinYin  Yes you are right.

**Comment by majianpeng** [comment]
> We have actually hit this same exact problem with krbd and went for a crude workaround in userspace -- retrying EBUSY after a short sleep.
Yes, i try use strace to debug. But it can't reproduce.  Only recompile kernel w/ more message to debug.

**Comment by ideepika** [comment]
> > @idryomov What's your opinion？
> 
> Your reasoning makes sense to me. I think we can close this PR in favor of #43461 but please carry over comments [#42775 (comment)](https://github.com/ceph/ceph/pull/42775#issuecomment-930688519), [#42775 (comment)](https://github.com/ceph/ceph/pull/42775#issuecomment-938360082) and [#42775 (comment)](https://github.com/ceph/ceph/pull/42775#issuecomment-938363753) into "librbd/cache/pwl: fix external flush dispatch in advance" commit message so that the trail of thought is recorded in the git history.

@CongMinYin @majianpeng should we close this PR in favor of https://github.com/ceph/ceph/pull/43461 ?

**Comment by CongMinYin** [comment]
> @CongMinYin @majianpeng should we close this PR in favor of #43461 ?

Yes, we should close this PR.

**Pull request closed without merging**
